cloud/scope/vpc_cluster.go

```diff
@@ -686,3 +742,321 @@ func (s *VPCClusterScope) buildCOSObjectHRef() (*string, error) {
+	} else {
+		s.SetResourceStatus(infrav1beta2.ResourceTypeWorkerSubnet, resourceStatus)
+	}
+	return requeue, nil
+}
+
+// createSubnet creates a new VPC subnet.
+func (s *VPCClusterScope) createSubnet(subnet infrav1beta2.Subnet, isControlPlane bool) error {
+	// Created resources should be placed in the cluster Resource Group (not Network, if it exists).
+	resourceGroupID, err := s.GetResourceGroupID()
+	if err != nil {
+		return fmt.Errorf("error retrieving resource group id for subnet creation: %w", err)
+	} else if resourceGroupID == "" {
+		return fmt.Errorf("error retrieving resource group id for resource group %s", s.IBMVPCCluster.Spec.ResourceGroup)
+	}
+
+	vpcID, err := s.GetVPCID()
+	if err != nil {
+		return fmt.Errorf("error retrieving vpc id for subnet creation: %w", err)
+	}
+
+	// TODO(cjschaef): Move to webhook validation.
+	if subnet.Zone == nil {
+		return fmt.Errorf("error subnet zone must be defined for subnet %s", *subnet.Name)
+	}
+
+	// NOTE(cjschaef): We likely will want to add support to use custom Address Prefixes
+	// For now, we rely on the API to assign us prefixes, as we request via IP count
+	var ipCount int64 = 256
+	// We currnetly only support IPv4
+	ipVersion := "ipv4"
+
+	// Find or create a Public Gateway in this zone for the subnet, only one Public Gateway is required for each zone, for this cluster.
+	// NOTE(cjschaef): We may need to add support to not attach Public Gateways to subnets.
+	publicGateway, err := s.findOrCreatePublicGateway(*subnet.Zone)
+	if err != nil {
+		return fmt.Errorf("error failed to find or create public gateway for subnet %s: %w", *subnet.Name, err)
+	}
+
+	options := &vpcv1.CreateSubnetOptions{}
+	options.SetSubnetPrototype(&vpcv1.SubnetPrototype{
+		IPVersion:             ptr.To(ipVersion),
+		TotalIpv4AddressCount: ptr.To(ipCount),
+		Name:                  subnet.Name,
+		VPC: &vpcv1.VPCIdentity{
+			ID: vpcID,
+		},
+		Zone: &vpcv1.ZoneIdentity{
+			Name: subnet.Zone,
+		},
+		ResourceGroup: &vpcv1.ResourceGroupIdentity{
+			ID: ptr.To(resourceGroupID),
+		},
+		PublicGateway: &vpcv1.PublicGatewayIdentity{
+			ID: publicGateway.ID,
+		},
+	})
+
+	// Create subnet.
+	subnetDetails, _, err := s.VPCClient.CreateSubnet(options)
+	if err != nil {
+		return fmt.Errorf("error unknown failure creating vpc subnet: %w", err)
+	}
+	if subnetDetails == nil || subnetDetails.ID == nil || subnetDetails.CRN == nil {
+		return fmt.Errorf("error failed creating subnet: %s", *subnet.Name)
+	}
+
+	// Initially populate subnet's status.
+	resourceStatus := &infrav1beta2.ResourceStatus{
+		ID:    *subnetDetails.ID,
+		Name:  subnetDetails.Name,
+		Ready: false,
+	}
+	if isControlPlane {
+		s.SetResourceStatus(infrav1beta2.ResourceTypeControlPlaneSubnet, resourceStatus)
+	} else {
+		s.SetResourceStatus(infrav1beta2.ResourceTypeWorkerSubnet, resourceStatus)
+	}
+
+	// Add a tag to the subnet for the cluster.
+	err = s.TagResource(s.IBMVPCCluster.Name, *subnetDetails.CRN)
```

Comment:
Hope tagging subnet is tested? We have nt tried before.

Reply:
Yes it does. I don't perform exhaustive testing, due to time constraints, but I test all the reconciliation logic after each set of changes for the standard path. So tagging works on subnets (based on the API I don't think GlobalTagging cares what a resource is, as long as it has a CRN), even though the UI leaves a lot to be desired in terms of tagging support.
```
# ibmcloud resource search "tags:\"us-east-capi-subnet-1-v2vzf\"" | grep "Name\|Resource Type"
Name:                us-east-capi-subnet-1-v2vzf-pgateway-us-east-1
Resource Type:       public-gateway
Name:                us-east-capi-subnet-1-v2vzf-subnet-compute-us-east-1
Resource Type:       subnet
Name:                us-east-capi-subnet-1-v2vzf-subnet-control-plane-us-east-1
Resource Type:       subnet
Name:                us-east-capi-subnet-1-v2vzf-pgateway-us-east-2
Resource Type:       public-gateway
Name:                us-east-capi-subnet-1-v2vzf-subnet-control-plane-us-east-2
Resource Type:       subnet
Name:                us-east-capi-subnet-1-v2vzf-subnet-compute-us-east-2
Resource Type:       subnet
Name:                us-east-capi-subnet-1-v2vzf-pgateway-us-east-3
Resource Type:       public-gateway
Name:                us-east-capi-subnet-1-v2vzf-subnet-compute-us-east-3
Resource Type:       subnet
Name:                us-east-capi-subnet-1-v2vzf-subnet-control-plane-us-east-3
Resource Type:       subnet
Name:                us-east-capi-subnet-1-v2vzf-rhcos
Resource Type:       image
Name:                us-east-capi-subnet-1-v2vzf-vpc
Resource Type:       vpc
```